core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1763,12 +1763,15 @@ private[zio] trait ZIOFunctions extends Serializable {
   final def foreachParN[R, E, A, B](
     n: Long
   )(as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    (for {
-      semaphore <- Semaphore.make(n)
-      bs <- foreachPar[R, E, A, B](as) { a =>
-             semaphore.withPermit(fn(a))
-           }
-    } yield bs).refailWithTrace
+    for {
+      q     <- Queue.bounded[(Promise[E, B], A)](n.toInt)
+      pairs <- ZIO.foreach(as)(a => Promise.make[E, B].map(p => (p, a)))
+      _     <- ZIO.foreach(pairs)(pair => q.offer(pair)).fork
+      _ <- ZIO.collectAll(List.fill(n.toInt)(q.take.flatMap {
+            case (p, a) => fn(a).foldCauseM(c => ZIO.foreach(pairs)(_._1.halt(c)), b => p.succeed(b))
+          }.forever.fork))
+      res <- ZIO.foreach(pairs)(_._1.await).ensuring(q.shutdown)
+    } yield res
```

Comment:
Gotta keep `refailWithTrace`, so that an error in child fiber points to the location of `foreachParN`. If you could also add a test for foreachParN similar to "foreachPar fail" in StacktracesSpec to prevent future regressions, that'd be great! 👍

Reply:
This is fixed now and I added a test. Thanks for your feedback on this!